File: chronicle-pc/frontend/src/notes/frontmatter.ts
const FM_RE = /^---\r?\n([\s\S]*?)\r?\n---/

const DEFAULT_KEY_ORDER = ['title', 'created', 'updated', 'type', 'tags', 'aliases'] as const

/** Parse simple YAML frontmatter (key: value lines). */
export function parseFrontmatter(content: string): {
  frontmatter: Record<string, string>
  body: string
} {
  const m = content.match(FM_RE)
  if (!m) return { frontmatter: {}, body: content }
  const fm: Record<string, string> = {}
  for (const line of m[1].split('\n')) {
    const idx = line.indexOf(':')
    if (idx <= 0) continue
    fm[line.slice(0, idx).trim()] = line.slice(idx + 1).trim()
  }
  return { frontmatter: fm, body: content.slice(m[0].length) }
}

function formatFrontmatter(fm: Record<string, string>): string {
  const ordered: [string, string][] = []
  const seen = new Set<string>()
  for (const key of DEFAULT_KEY_ORDER) {
    if (key in fm) {
      ordered.push([key, fm[key]])
      seen.add(key)
    }
  }
  for (const [k, v] of Object.entries(fm)) {
    if (!seen.has(k)) ordered.push([k, v])
  }
  const lines = ordered.map(([k, v]) => `${k}: ${v}`)
  return `---\n${lines.join('\n')}\n---`
}

function setKey(
  fm: Record<string, string>,
  key: string,
  value: string,
  overwrite: boolean,
): void {
  const lowerMap = new Map(Object.keys(fm).map((k) => [k.toLowerCase(), k]))
  const existing = lowerMap.get(key.toLowerCase())
  if (existing !== undefined) {
    if (!overwrite && fm[existing].trim() !== '') return
    if (existing !== key) delete fm[existing]
  }
  fm[key] = value
}

/** Set or add ``updated:`` to today's ISO date (YYYY-MM-DD). */
export function bumpUpdated(content: string, today = new Date().toISOString().slice(0, 10)): string {
  const { frontmatter, body } = parseFrontmatter(content)
  frontmatter.updated = today
  return `${formatFrontmatter(frontmatter)}${body}`
}

/**
 * Fill missing create-time frontmatter (created/updated/type/title/tags)
 * per `_system/conventions.md`. Existing keys are preserved; ``updated`` is stamped.
 */
export function ensureCreateFrontmatter(
  content: string,
  opts: { title?: string; type?: string; today?: string } = {},
): string {
  const today = opts.today ?? new Date().toISOString().slice(0, 10)
  const { frontmatter, body } = parseFrontmatter(content)
  const title = opts.title?.trim()
  if (title) setKey(frontmatter, 'title', title, false)
  setKey(frontmatter, 'created', today, false)
  setKey(frontmatter, 'updated', today, true)
  setKey(frontmatter, 'type', opts.type ?? 'note', false)
  setKey(frontmatter, 'tags', '[]', false)
  let nextBody = body
  if (nextBody && !nextBody.startsWith('\n')) nextBody = `\n${nextBody}`
  if (!nextBody) nextBody = '\n'
  return `${formatFrontmatter(frontmatter)}${nextBody}`
}

File: chronicle-pc/server/src/frontmatter.rs
//! Frontmatter parsing/completion (note_frontmatter.py) + stable renderer.

use std::collections::BTreeMap;

const DEFAULT_KEY_ORDER: [&str; 6] = ["title", "created", "updated", "type", "tags", "aliases"];

pub struct ParsedFrontmatter {
    pub fm: BTreeMap<String, String>,
    pub body: String,
}

/// \A---\r?\n(.*?)\r?\n---\r?\n? — DOTALL, non-greedy.
pub fn parse_frontmatter(text: &str) -> ParsedFrontmatter {
    let bytes = text.as_bytes();
    if !text.starts_with("---") {
        return ParsedFrontmatter { fm: BTreeMap::new(), body: text.to_string() };
    }
    // Find end of first line.
    let after_open = match find_line_end(bytes, 3) {
        Some(idx) => idx,
        None => return ParsedFrontmatter { fm: BTreeMap::new(), body: text.to_string() },
    };
    // Find closing fence line "---".
    let mut search = after_open;
    while let Some(rel) = text[search..].find("\n---") {
        let abs = search + rel;
        let line_start = abs + 1;
        let rest = &text[line_start..];
        let is_fence = rest.starts_with("---")
            && rest[3..].starts_with(|c: char| c == '\r' || c == '\n' || c == '\0')
            || rest == "---";
        if is_fence {
            let after_fence = line_start + 3;
            let body_start = skip_one_newline(bytes, after_fence);
            let inner = &text[after_open..abs];
            let mut fm = BTreeMap::new();
            for line in inner.lines() {
                if let Some((k, v)) = line.split_once(':') {
                    fm.insert(k.trim().to_string(), v.trim().to_string());
                }
            }
            return ParsedFrontmatter { fm, body: text[body_start..].to_string() };
        }
        search = abs + 1;
    }
    ParsedFrontmatter { fm: BTreeMap::new(), body: text.to_string() }
}

fn find_line_end(bytes: &[u8], from: usize) -> Option<usize> {
    let mut i = from;
    while i < bytes.len() {
        if bytes[i] == b'\n' {
            return Some(i + 1);
        }
        i += 1;
    }
    None
}

fn skip_one_newline(bytes: &[u8], at: usize) -> usize {
    if at < bytes.len() && bytes[at] == b'\r' && at + 1 < bytes.len() && bytes[at + 1] == b'\n' {
        at + 2
    } else if at < bytes.len() && bytes[at] == b'\n' {
        at + 1
    } else {
        at
    }
}

fn set_key(fm: &mut BTreeMap<String, String>, key: &str, value: &str, overwrite: bool) {
    let existing = fm
        .keys()
        .find(|k| k.eq_ignore_ascii_case(key))
        .cloned();
    if let Some(cased) = existing {
        let non_empty = fm.get(&cased).map(|v| !v.trim().is_empty()).unwrap_or(false);
        if !overwrite && non_empty {
            return;
        }
        fm.remove(&cased);
    }
    fm.insert(key.to_string(), value.to_string());
}

pub fn format_frontmatter(fm: &BTreeMap<String, String>) -> String {
    let mut lines: Vec<String> = Vec::new();
    for key in DEFAULT_KEY_ORDER {
        if let Some(v) = fm.get(key) {
            lines.push(format!("{key}: {v}"));
        }
    }
    for (k, v) in fm {
        if !DEFAULT_KEY_ORDER.contains(&k.as_str()) {
            lines.push(format!("{k}: {v}"));
        }
    }
    format!("---\n{}\n---", lines.join("\n"))
}

pub fn today_iso() -> String {
    chrono::Local::now().format("%Y-%m-%d").to_string()
}

/// ensure_create_frontmatter — create-time completion contract.
pub fn ensure_create_frontmatter(
    content: &str,
    title: Option<&str>,
    note_type: &str,
) -> String {
    let parsed = parse_frontmatter(content);
    let mut fm = parsed.fm;
    if let Some(t) = title.map(str::trim).filter(|t| !t.is_empty()) {
        set_key(&mut fm, "title", t, false);
    }
    set_key(&mut fm, "created", &today_iso(), false);
    set_key(&mut fm, "updated", &today_iso(), true);
    set_key(&mut fm, "type", note_type, false);
    set_key(&mut fm, "tags", "[]", false);
    let mut body = parsed.body;
    if !body.is_empty() && !body.starts_with('\n') {
        body.insert(0, '\n');
    }
    if body.is_empty() {
        body.push('\n');
    }
    format!("{}{}", format_frontmatter(&fm), body)
}

/// _stable_frontmatter (notes.py): keys sorted alphabetically, newlines→spaces.
pub fn stable_frontmatter(pairs: &[(&str, String)]) -> String {
    let mut sorted: Vec<(String, String)> = pairs
        .iter()
        .map(|(k, v)| (k.to_string(), v.replace('\n', " ").trim().to_string()))
        .collect();
    sorted.sort_by(|a, b| a.0.cmp(&b.0));
    let lines: Vec<String> = sorted.iter().map(|(k, v)| format!("{k}: {v}")).collect();
    format!("---\n{}\n---", lines.join("\n"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn completes_create_frontmatter_in_order() {
        let out = ensure_create_frontmatter("Hello world", Some("My Note"), "note");
        assert!(out.starts_with("---\ntitle: My Note\ncreated: "));
        assert!(out.contains("\nupdated: "));
        assert!(out.contains("\ntype: note\ntags: []\n---\nHello world"));
    }

    #[test]
    fn respects_existing_nonempty_keys_case_insensitively() {
        let src = "---\nTitle: Kept\n---\nBody";
        let out = ensure_create_frontmatter(src, Some("New"), "note");
        assert!(out.contains("Title: Kept"), "{out}");
        assert!(!out.contains("title: New"));
    }

    #[test]
    fn stable_frontmatter_sorts_and_flattens_newlines() {
        let out = stable_frontmatter(&[("zeta", "1".into()), ("alpha", "a\nb".into())]);
        assert_eq!(out, "---\nalpha: a b\nzeta: 1\n---");
    }
}
